plutus-playground/plutus-playground-server/src/Playground/Interpreter.hs

```diff
@@ -1,166 +1,195 @@
-        setImportsQ
-            [ ("Playground.Interpreter.Util", Nothing)
-            , ("Playground.API", Nothing)
-            , ("Wallet.Emulator", Nothing)
-            , ("Ledger.Types", Nothing)
-            , ("Data.Map", Nothing)
-            ]
-        liftIO . putStrLn $ expr
-        res <-
-            interpret
-                expr
-                (as :: Either PlaygroundError ( Blockchain
-                                              , [EmulatorEvent]
-                                              , [(Wallet, Value)]))
-        case res of
-            Left e  -> throwM . UnknownError $ show e
-            Right r -> pure r
-
-mkExpr :: (MonadError PlaygroundError m) => Evaluation -> m String
+    withSystemTempFile "Main.hs" $ \file handle -> do
+        (exitCode, stdout, stderr) <-
+            runscript handle file $
+            mkRunScript (Newtype.unpack source) (Text.pack . BS8.unpack $ expr)
+        result <-
+            case exitCode of
+                ExitSuccess -> pure stdout
+                _           -> throwError . InterpreterError $ [stderr]
+        let decodeResult =
+                JSON.eitherDecodeStrict . BS8.pack $ result :: Either String (Either PlaygroundError ( Blockchain
+                                                                                                     , [EmulatorEvent]
+                                                                                                     , [( Wallet
+                                                                                                        , Value)]))
+        case decodeResult of
+            Left err ->
+                throwError . OtherError $
+                "unable to decode compilation result" <> err
+            Right eResult ->
+                case eResult of
+                    Left err      -> throwError err
+                    Right result' -> pure result'
+
+mkRunScript :: Text -> Text -> Text
+mkRunScript script expr =
+    replaceModuleName script <> "\n\nmain :: IO ()" <> "\nmain = printJson $ " <>
+    expr
+
+runghcOpts :: [String]
+runghcOpts =
+    [ "-XDataKinds"
+    , "-XDeriveAnyClass"
+    , "-XDeriveFoldable"
+    , "-XDeriveFunctor"
+    , "-XDeriveGeneric"
+    , "-XDeriveLift"
+    , "-XDeriveTraversable"
+    , "-XExplicitForAll"
+    , "-XFlexibleContexts"
+    , "-XOverloadedStrings"
+    , "-XRecordWildCards"
+    , "-XStandaloneDeriving"
+    , "-XTemplateHaskell"
+    , "-XScopedTypeVariables"
+    , "-O0"
+    ]
+
+lookupRunghc :: (MonadIO m, MonadError PlaygroundError m) => m String
+lookupRunghc = do
+    mBinDir <- liftIO $ lookupEnv "GHC_BIN_DIR"
+    case mBinDir of
+        Nothing  -> pure "runghc"
+        Just val -> pure $ val <> "/runghc"
+
+{-# ANN ignoringIOErrors ("HLint: ignore Evaluate" :: String) #-}
+
+ignoringIOErrors :: MonadCatch m => m () -> m ()
+ignoringIOErrors ioe = ioe `catch` (\e -> const (return ()) (e :: IOError))
+
+withSystemTempFile ::
```

Comment:
I'm surprise this isn't in a lib somewhere...

Reply:
unfortunately I needed to remove the concrete `IO` monad and that is pretty difficult, probably can be done with unliftIO or something like that but it's tricky. Seemed easier just to copy and paste from the temp library.

Reply:
Then a comment explaining that would be good!